q2_coordinates/qtrees.py

```diff
@@ -0,0 +1,204 @@
+        return self.points
+    
+    def get_id(self):
+        return self.id
+    
+    def set_id(self, _id):
+        self.id = self.id+_id
+
+    
+class QTree():
+    def __init__(self, k, data):
+        self.threshold = k
+        self.points = [Point(x, y, sample_id) for sample_id, x, y in data]
+        self.root = Node(0, 0, 360, 180, self.points, "0")
+
+    def add_point(x, y, sample_id):
+        self.points.append(Point(x, y, sample_id))
+    
+    def get_points(self):
+        return self.points
+    
+    def subdivide(self, samples, threshold):
+        count = 0
+        node_id = ""
+        samples = samples
+        bins  = []
+        
+        samples, bins = recursive_subdivide(self.root, threshold, count, node_id, samples, bins)
+        return samples, bins
+    
+
+def recursive_subdivide(node, k, count, node_id, samples, bins):
+
+    if len(node.points) < k:
+        return
+    elif len(node.points)/k >= len(node.points):
+        raise ValueError("The threshold for subdivision is less than the amount of points in one place,", 
+                         "please chose a larger threshold for division")
+    else:
+        count += 1
+
+    samples["H" + str(count)] = ""
+    
+    w_ = node.width/2
+    h_ = node.height/2    
+    
+    nodes = []
+    sk_nodes = []
+    for i in range(4):
+        #southwest
+        if(i ==0):
+            quad = "3"
+            node.x0 = node.x0
+            node.y0 = node.y0
+        #northwest
+        elif(i==1):
+            quad = "1"
+            node.x0 = node.x0
+            node.y0 = node.y0+h_
+        #northeast
+        elif(i == 2):
+            quad = "2"
+            node.x0 = node.x0+w_
+            node.y0 = node.y0
+        #southeast
+        elif(i == 3):
+            quad = "4"
+            node.x0 = node.x0
+            node.y0 = node.y0-h_
+
+        p = contains(node.x0, node.y0, w_, h_, node.points)
+        quad_node= Node(node.x0, node.y0, w_, h_, p, node_id+quad+".")
+        
+        for pt in p:
+            bins.append((pt.sample_id, count, quad_node.get_id()))
+        
+        nodes.append(quad_node)
+        recursive_subdivide(quad_node, k, count, node_id+quad+".", samples, bins)
+        
+
+    node.children = nodes
+    return samples, bins
+
+    #if number of input samples/k is less than 4 there may be issues raise exception
+    
+    
+def contains(x, y, w, h, points):
+    pts = []
+    for point in points:
+        if point.x >= x and point.x <= x+w and point.y>=y and point.y<=y+h:
+            pts.append(point)
+    return pts
+
+def create_tree(samples):
+    tree = skbio.TreeNode(name="root")
+    lineages = []
+    for i in samples.iterrows():
+        values = list(i[1])
+        
+        lineages.append((i[0], values))
+
+    tree.extend(skbio.TreeNode.from_taxonomy(lineages))
+    return tree
+
+def create_sample_df(bins, samples, index):
+    print(samples.head())
+    print(index)
+    for samp, bin_i, items in bins:
```

Comment:
I think this might be able to be expressed as:

```python
bins = pd.DataFrame(bins, names=['sample-id', 'bin-index', 'lineage']).set_index('sample-id')
bins['bin-name'] = ["H" + str(i) for i in bins['bin-index']]
return bins
```